### PR TITLE
[oracle]Use Incremental Snapshot Framework for Oracle CDC Connector

### PR DESCRIPTION
**Pull request by molsionmo**
[oracle]Use Incremental Snapshot Framework for Oracle CDC Connector #1079  @leonardBang 
1. flink-connector-oracle-cdc incremental Snapshot implemention base on flink-cdc-base
2. It lacks tests, and then replenishes the corresponding test classes and fixes the problems found by the tests
3. Below is the file directory and the implementation file description
- source
  - assigner
    - splitter
      - OracleChunkSplitter.java(used to split table into a set of chunks for oracle data source)
  - config
    - OracleSourceConfig.java
    - OracleSourceConfigFactory.java
    - OracleSourceOptions.java
  - meta
    - offset
      - RedoLogOffset.java(A structure describes an offset in a redo log event in oracle)
      - RedoLogOffsetFactory.java
      - RedoLogOffsetSerializer.java
    - split
      - FinishedSnapshotSplitInfo.java(The information used to describe a finished snapshot split)
      - OracleRedoLogSplit.java(The split to describe the redo log of Oracle table(s))
      - OracleSnapshotSplit.java(The split to describe a split of an Oracle table snapshot)
      - OracleSplit.java(The split of table comes from a Table that splits by primary key.)
  - reader
    - fetch
      - OracleScanFetchTask.java(The task to work for fetching data of Oracle table snapshot split.)
      - OracleSourceFetchTaskContext.java
      - OracleStreamFetchTask.java(The task to work for fetching data of Oracle table stream split)
  - utils
    - OracleConnectionUtils.java
    - OracleSchema.java
    - OracleTypeUtils.java
    - OracleUtils.java
  - EmbeddedFlinkDatabaseHistory.java(A {@link DatabaseHistory} implementation which store the latest table schema in Flink state.)
  - OracleDialect.java(The {@link JdbcDataSourceDialect} implementation for Oracle datasource)
  - OraclePooledDataSourceFactory
  - OracleSourceBuilder

- debezium
       - logminer
         - LogMinerStreamingChangeEventSource(copy from io.debezium.connector.oracle.logminer.LogMinerStreamingChangeEventSource,and add afterHandleScn method to handle scn event, just like io.debezium.connector.mysql.MySqlStreamingChangeEventSource#handleEvent)
         - TransactionalBuffer
       - OracleErrorHandler

### CODE REVIEW

**Comment by molsionmo** [comment]
These is a question, in OracleStreamFetchTask.java , LogMinerStreamingChangeEventSource does not have handle(event) method so that we can not stop for fetch binlog for snapshot split. I can not kown how to resolve this question.

I have a solution that we can reimplements StreamingChangeEventSource that copy from LogMinerStreamingChangeEventSource and add event handle method, but i do not know if it works

**Comment by molsionmo** [comment]
I run OracleE2eITCase in my compute successfully, i guess it is time out after 150 seconds

proxy.checkResultWithTimeout(
                expectResult,
                "products_sink",
                new String[] {"id", "name", "description", "weight"},
                150000L);

**Comment by fuyun2024** [comment]
> 这是一个问题，在 OracleStreamFetchTask.java 中，LogMinerStreamingChangeEventSource 没有 handle(event) 方法，因此我们无法停止获取 binlog 以进行快照拆分。我不知道如何解决这个问题。
> 
> 我有一个解决方案，我们可以重新实现从 LogMinerStreamingChangeEventSource 复制的 StreamingChangeEventSource 并添加事件句柄方法，但我不知道它是否有效@leonardBang你能给我一些建议吗？谢谢

hi , @molsionmo  我之前调研的时候做了一个简单的验证，你可以参考一下：
1、重写 LogMinerStreamingChangeEventSource#execute   方法，把原本在 execute 方法中创建 TransactionalBuffer 的操作，改成可以右外部输入  setTransactionalBuffer ()。

2、继承 transactionalBuffer 类，实现 commit() 方法，在此方法中获取到消费的位置。

这个是我之前的验证过程，它能获取到消费的位置，但是实现方式不算是很优雅。

**Comment by molsionmo** [comment]
Hi @fuyun2024  Thanks for your comment. You inspired me. I did it with a simpler way.

OracleRedoLogSplitReadTask implement afterHandleScn method and check do we need to stop for fetch redo log for snapshot split.OracleLogMinerStreamingChangeEventSource copy from LogMinerStreamingChangeEventSource and add  afterHandleScn() that can handle it after handle scn

**Comment by fuyun2024** [comment]
Hi @molsionmo I'm glad you have a simpler way,  but I think it has some optimization. The afterHandleScn () method is called after processResult (), so scn in offsetContext will cross many commit.

**Comment by leonardBang** [comment]
I open a PR to avoid duplicate records when restoration #1100, maybe my fix way can inspire you a a little @molsionmo @fuyun2024

**Comment by molsionmo** [comment]
@leonardBang @GOODBOY008 @fuyun2024 I'm excited to tell you that I have passed the test. I'll merge the code, including the test code, on Sunday. I'll add end-to-end testing later

**Review by molsionmo — COMMENTED** [review]
Thank you for your good job!
1. If add RedoLogOffsetTest will be great
2. [OracleSourceSplitSerializer.java][SerializerUtils.java]  i think should be removed beause they are annotation all
3. Is there a problem with removing the serialized class in sense of checkpoint? Please see com.ververica.cdc.connectors.oracle.source.OracleSourceITCase#testTaskManagerFailoverInSnapshotPhase; com.ververica.cdc.connectors.oracle.source.OracleSourceITCase#testTaskManagerFailoverInBinlogPhase
4. if you can commit after run (mvn spotless:apply  & mvn clean package -Dmaven.javadoc.skip=true -DskipTests=true) will be greater

File: flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/OracleSourceBuilder.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The builder class for {@link OracleIncrementalSource} to make it easier for the users to
+ * construct a {@link OracleIncrementalSource}.
+ *
+ * <p>Check the Java docs of each individual method to learn more about the settings to build a
+ * {@link OracleIncrementalSource}.
+ */
+@Experimental
```
Comment (leonardBang) [inline flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/OracleSourceBuilder.java:41]: hint: `@Internal`

File: flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/meta/split/OracleSnapshotSplit.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import io.debezium.relational.history.TableChanges;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/** The split to describe a split of an Oracle table snapshot. */
```
Comment (leonardBang) [inline flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/meta/split/OracleSnapshotSplit.java:34]: We do not need any Split as the flink-cdc-base has done these work

File: flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/meta/split/OracleSplit.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.flink.api.connector.source.SourceSplit;
+
+import io.debezium.relational.TableId;
+import io.debezium.relational.history.TableChanges;
+
+import java.util.Map;
+import java.util.Objects;
+
+/** The split of table comes from a Table that splits by primary key. */
+public abstract class OracleSplit implements SourceSplit {
```
Comment (leonardBang) [inline flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/meta/split/OracleSplit.java:30]: We do not need any Split as the flink-cdc-base has done these work

**Review by leonardBang — COMMENTED** [review]
Thanks @molsionmo and @fuyun2024 for the huge work, I left some comments

File: flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/EmbeddedFlinkDatabaseHistory.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * A {@link DatabaseHistory} implementation which store the latest table schema in Flink state.
+ *
+ * <p>It stores/recovers history using data offered by {@link SourceSplitState}.
+ */
+public class EmbeddedFlinkDatabaseHistory implements DatabaseHistory {
```
[the remark this answers is not in the document]
Comment (molsionmo) [reply flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/source/EmbeddedFlinkDatabaseHistory.java:46]: I have moved this class to com.ververica.cdc.connectors.base.source.EmbeddedFlinkDatabaseHistory

**Review by leonardBang — APPROVED** [review]
Thanks @molsionmo and @fuyun2024 for the great work, LGTM

File: flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/util/ChunkUtils.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            throw new ValidationException(
+                    String.format(
+                            "Chunk key column '%s' doesn't exist in the primary key [%s] of the table %s.",
+                            chunkKeyColumn,
+                            primaryKeys.stream().map(Column::name).collect(Collectors.joining(",")),
+                            table.id()));
+        }
+
+        // Use the ROWID column as the chunk key column by default for oracle cdc connector
```
Comment (e-mhui) [inline flink-connector-oracle-cdc/src/main/java/com/ververica/cdc/connectors/oracle/util/ChunkUtils.java:56]: @molsionmo Hi，if there's a primary key, why not use it as the chunk key?

**Comment by loserwang1024** [comment]
>  add afterHandleScn() that can handle it after handle scn.

@molsionmo , it seems some problem will occur(See https://github.com/ververica/flink-cdc-connectors/issues/2909).

> LogMinerStreamingChangeEventSource does not have handle(event) method so that we can not stop for fetch binlog for snapshot split. I can not kown how to resolve this question.

Almost all subclass of `AbstractLogMinerEventProcessor` provide a `processRow` method, I try in https://github.com/ververica/flink-cdc-connectors/pull/2911, it works.

**Comment by tttzzzwww** [comment]
hello！In flink cdc 2.2, flink does need to build table permission and write permission, because in io.debezium.connector. oracle.logminer Log location of the latest read to be logged, so a LOG _ MINING _ FLUSH was created to store the location of the most recent read log. But I think I have fixed this problem in flink cdc 2.3.https://github.com/apache/flink-cdc/commit/dd7b7beaaa39095202b7fb0957b32fae43bdbfc6#diff-fe1123b753e8883b6e03a1fe6c469e234c3c00de803e5be2ba6b098d9f6b4257 and https: / / github. com/apache/flink-cdc/commit/70fb9cf5a3fe810f5d9b9d66f9fa251db2dfd678#diff-fe1123b753e8883b6e03a1fe6c469e234c3c00de803e5be2ba6b098d9f6b4257
However, the debezium version was later upgraded in version 2.4, and the operation of creating the LOG _ MINING _ FLUSH table and updating the table was added back
example：
image
**Want to consult below, for the source library, build table permission is neede**

**Pull request merged**